jep/0000/README.adoc

```diff
@@ -0,0 +1,345 @@
+** `autoreleaseEnabled`: A boolean value, indicating whether or not this plugin is participating 
+in autorelease. Allows the maintainer to turn Autorelease on or off, without deleting the file 
+entirely.
+** `triggeringBranch`: Sets the branch which maintainers will merge to in order to trigger 
+an autorelease webhook. Although this is expected to usually be set to `master`, plugin 
+maintainers can choose a branch name of their preference, e.g., `autorelease`, `release`, 
+etc.
+** `versionNumberFields`: An integer, indicating the number of fields which `buildPlugin` should expect.
+This is important, because sticking to a consistent numbering scheme makes for easier sorting by version.
+For this example plugin, Karl has set the value to `3`, because he's usuing semantic versioning, e.g., 
+3.1.4.
+
+==== Maven POM file changes
+
+Karl's plugin is already making use of Incrementals, and the relevant lines of his `pom.xml` file 
+look like:
+
+```
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.jenkins-ci.plugins</groupId>
+        <artifactId>plugin</artifactId>
+        <version>4.56</version>
+        <relativePath />
+    </parent>
+    <artifactId>karl</artifactId>
+    <version>${revision}${changelist}</version>
+    <packaging>hpi</packaging>
+    <name>Karl Plugin</name>
+    <url>
+        <!--Something like https://wiki.jenkins-ci.org/display/JENKINS/Karl+Plugin-->
+    </url>
+    <description>A useful description of the Karl plugin.</description>
+    <licenses>
+        <license>
+            <name>MIT</name>
+            <url>http://opensource.org/licenses/MIT</url>
+        </license>
+    </licenses>
+
+    <properties>
+        <revision>3.1.4</revision>
+        <changelist>-SNAPSHOT</changelist>
+        <java.level>8</java.level>
+        <jenkins.version>2.138.4</jenkins.version>
+    </properties>
+
+```
+
+No additional changes need to be made to `pom.xml` by Karl, he's good to go.
+
+==== Autorelease.yaml verification in `buildPlugin`
+
+The commonly used library link:https://github.com/jenkins-infra/pipeline-library/blob/master/vars/buildPlugin.groovy[`buildPlugin`] will need to be modified to check for the presence of, and validate, `Autorelease.yaml`. This 
+markerfile will be the mechanism that tells `${trusted-ci-server}` that this plugin should be automatically 
+released. 
+
+Validation must include, but will not be limited to, the following:
+
+1. `pluginName` field must match the repository name
+
+2. `versionNumberFields` must match the `revision` property in `pom.xml`. E.g., if `versionNumberFields` is 
+set to `3` (three fields), and the `revision` in `pom.xml` is set to `3.1.4.5` (four fields), validation 
+will fail.
```

Comment:
So…why require it to be specified to begin with?

Reply:
This is related to something that @daniel-beck and I talked about. I understood it in the moment, but the passage of time means I've lost the plot a little. We wanted the number of fields to remain consistent within a given plugin, to avoid having versions sort incorrectly.

Reply:
If the pom declares a version prefix that is later amended by the tool, say, `3.1` or `3.1.${build}` or equivalent, we may run into the problem that a "later looking" version number is actually sorting as earlier.

Imagine the fixed version prefix `3.1`, and autorelease adding a build number:

3.1.0
3.1.1
3.1.2
3.1.3

Now, the maintainer changes the apparent semver to `3.1.1`.

3.1.1.0 (or 3.1.1.4, doesn't matter)
3.1.1.1
…

While these releases should be later, they sort earlier. Fixed length version numbers (or "consistent length" per component) addresses this problem.

I also noted this might be circumventable by using a different scheme here – E.g. if something like `3.1.1-b4` or similar would sort differently.

Reply:
> Fixed length version numbers addresses this problem.

Until the maintainer

```diff
-versionNumberFields: 2
+versionNumberFields: 3
```

to fix a build error, without considering the consequences on the update center.

> something like `3.1.1-b4` or similar would sort differently

As per [JEP-305](https://github.com/jenkinsci/jep/blob/master/jep/305/README.adoc#other-version-schemes), which is not susceptible to this problem because the generated versions always sort _before_ the maintainer-specified string. Of course that particular format is undesirable in this context since these are not “release candidates”, these are full releases—there is no other.

----

My suggestion would be to just get rid of the maintainer-specified component altogether in an autoreleased plugin. It serves no function beyond [pretending to indicate compatibility](https://twitter.com/tyvole/status/1136767636602327042), which is more helpfully and honestly accomplished by writing good release notes. With a minor enhancement to `git-changelist-maven-extension` to allow the `-rc` component to be skipped, the POM can just say

```xml
<version>${changelist}</version>
<properties>
  <changelist>999999-SNAPSHOT</changelist>
</properties>
```

so development builds are all `999999-SNAPSHOT` and incremental builds (including `master` releases) look like `1234.abc123def456`. The more changes you make, the bigger the number, and you never have to pay attention to the version again.

Reply:
…in which case we do not need another server, or even a new YAML file. ci.jenkins.io will already publish these releases. We would “just” need to patch `update-center2` to also check the Incrementals repository and include the newest release it finds there which

* Matches `\d+[.][a-f0-9]{12}` (i.e., you have opted in by virtue of dropping the manual `revision`).
* Comes from `master` according to [the same logic already used by `incrementals:update`](https://github.com/jenkinsci/incrementals-tools/blob/9a96211ad69fa55599001c9689b5c7e78f9a7f88/lib/src/main/java/io/jenkins/tools/incrementals/lib/UpdateChecker.java#L141-L146).

Reply:
> patch `update-center2` to also check the Incrementals repository

Better yet, make the `incrementals-publisher` function send artifacts matching these criteria to the regular `releases` repository; or have a separate function or batch process which copies from `incrementals` to `releases`. Thus

* These releases may be made dependencies of other plugins without `-Pconsume-incrementals`.
* We can experiment with GC policies in `incrementals` without fear of dropping bits which were at some point offered on the regular user UC.
* `update-center2` need not be touched—these artifacts would appear in the `releases` index and be published automatically.

Reply:
> to fix a build error, without considering the consequences on the update center.

From that point on, all autoreleases will necessarily need to fail.

> get rid of the maintainer-specified component altogether in an autoreleased plugin

I expect this will severely limit the acceptance of this system. Not everyone is comfortable with a version number "2412", especially if Autorelease is just being tried out for a given plugin -- no way to get back to version 3.0.x or whatever, after you've released version 2412.

Another problem would be projects like library plugins keeping lockstep with upstream releases, just adding another digit for the Jenkins plugin itself.

Reply:
> From that point on, all autoreleases will necessarily need to fail.

… unless accompanied by a major version increase. This is something that a PR build status could tell maintainers.

----

I think we'll have the best chance by imposing as few unnecessary restrictions as on plugin developers, if we want it to gain acceptance. Making the branch and the version scheme configurable is one of those things that should be doable without crazy workarounds and will make it easier for people to give this a try.

Reply:
> projects like library plugins keeping lockstep with upstream releases, just adding another digit for the Jenkins plugin itself

```xml
<artifactId>apache-httpcomponents-client-4-api</artifactId>
<version>4.5.5.${changelist}</version>
<properties>
  <changelist>999999-SNAPSHOT</changelist>
</properties>
```

producing versions such as `4.5.5.66.b6af896b0883`, which will sort reasonably. Does mean you would need to pick another flag for `incrementals-publisher`.

Reply:
Then the library releases version `4.5.5.1`. What now?

Reply:
What is basically what this proposal comes down to, with an extra layer to ensure going from `1.2` to `1.2.1` will not have unintended consequences.

Reply:
> you would need to pick another flag for `incrementals-publisher`

For the record: a POM property would not work since the flattened POM omits properties, but you could check for the presence/contents of a named file at that commit.

> What is basically what this proposal comes down to

Really? The proposal as currently written sounds very different to me: a special webhook, a special CI server, etc. The approach I am discussing is something I could probably implement in about a day as a single patch to `incrementals-publisher` (plus prototype usage in some minor plugin I control), given the ability to test changes iteratively like I had from @rtyler when it was initially developed.

> an extra layer to ensure going from `1.2` to `1.2.1` will not have unintended consequences

Is that danger actually specific to this JEP or autoreleasing generally? You can already produce shadowed releases with `maven-release-plugin`, admittedly via a somewhat different kind of source edit. Maybe we should consider other ways of dealing with this class of mistake more broadly, such as having `update-center2` produce a warning (delivered somewhere TBD) when it notices a chronologically new release which is not going to actually be visible in any of the configured update centers. (Not _necessarily_ a mistake, since the maintainer might be producing this release for consumption in some other channel, such as an OEM WAR.)

Reply:
@kshultzCB 
This is an example of a discussion that is non-blocking for this first PR.  You'll want to make sure that the salient points are added to the reasoning section, but this is discussion of the design itself not of the structure of the JEP.

Reply:
> you would need to pick another flag for `incrementals-publisher`

Oh! No special flag needed. Only that `-rc` is not present in the version number.

Reply:
```suggestion
will fail.

[WARNING]
====
*(bitwiseman)* 
There's a long thread here: https://github.com/jenkinsci/jep/pull/244/files#r293534268 that needs to be integrated into this. 





```